test/weiner_convolve_test.cc

```diff
@@ -0,0 +1,378 @@
+
+  protected:
+    SVTRandom rnd8_;
+    SVTRandom rnd16_;
+};
+
+static const int w = 128, h = 128;
+
+typedef void (*WienerConvolveFunc)(const uint8_t* src, ptrdiff_t src_stride,
+                                   uint8_t* dst, ptrdiff_t dst_stride,
+                                   const int16_t* filter_x, int x_step_q4,
+                                   const int16_t* filter_y, int y_step_q4,
+                                   int w, int h,
+                                   const ConvolveParams* conv_params);
+typedef void (*HbdWienerConvolveFunc)(const uint8_t* src, ptrdiff_t src_stride,
+                                      uint8_t* dst, ptrdiff_t dst_stride,
+                                      const int16_t* filter_x, int x_step_q4,
+                                      const int16_t* filter_y, int y_step_q4,
+                                      int w, int h,
+                                      const ConvolveParams* conv_params,
+                                      int32_t bps);
+typedef std::tuple<int, int, WienerConvolveFunc> WienerConvolveParam;
+typedef std::tuple<int, int, HbdWienerConvolveFunc, int> HbdWienerConvolveParam;
+
+template <typename Sample, typename FuncType, typename ParamType>
+class AV1WienerConvolveTest : public ::testing::TestWithParam<ParamType>,
+                              public SVTRandomHelper {
+  public:
+    void SetUp() override {
+        input_ = nullptr;
+        output_tst_ = nullptr;
+        output_ref_ = nullptr;
+        out_w_ = 0;
+        out_h_ = 0;
+        func_tst_ = nullptr;
+        bd_ = 8;
+        rnd8_.reset();
+        rnd16_.reset();
+        prepare_data();
+    }
+    void TearDown() override {
+        if (input_) {
+            delete[] input_;
+            input_ = nullptr;
+        }
+        if (output_tst_) {
+            delete[] output_tst_;
+            output_tst_ = nullptr;
+        }
+        if (output_ref_) {
+            delete[] output_ref_;
+            output_ref_ = nullptr;
+        }
+        aom_clear_system_state();
+    }
+
+  protected:
+    void prepare_data() {
+        input_ = new Sample[w * h];
+
+        // The AVX2 convolve functions always write rows with widths that are
+        // multiples of 16. So to avoid a buffer overflow, we may need to pad
+        // rows to a multiple of 16.
+        int output_n = ALIGN_POWER_OF_TWO(out_w_, 4) * out_h_;
+        output_tst_ = new Sample[output_n];
+        output_ref_ = new Sample[output_n];
+    }
+    virtual void run_check_output(const int test_times) = 0;
+
+  protected:
+    FuncType func_tst_;  /**< function to test */
+    uint32_t out_w_;     /**< width of output */
+    uint32_t out_h_;     /**< height of output */
+    uint32_t bd_;        /**< bit-depth */
+    Sample* input_;      /**< buffer of input data */
+    Sample* output_tst_; /**< buffer of test output data */
+    Sample* output_ref_; /**< buffer of reference output data */
+};
+
+class AV1WienerConvolveLbdTest
```

Comment:
Seems like LbdTest and HbdTest have lots in common, and could be simplified further.

Reply:
done